geopandas/tools/clip.py

```diff
@@ -0,0 +1,246 @@
+        raise ValueError("gdf and clip_obj extent do not overlap.")
+
+    if isinstance(clip_obj, (GeoDataFrame, GeoSeries)):
+        poly = clip_obj.geometry.unary_union
+    else:
+        poly = clip_obj
+
+    geom_types = gdf.geometry.type
+    poly_idx = np.asarray((geom_types == "Polygon") | (geom_types == "MultiPolygon"))
+    line_idx = np.asarray(
+        (geom_types == "LineString")
+        | (geom_types == "LinearRing")
+        | (geom_types == "MultiLineString")
+    )
+    point_idx = np.asarray((geom_types == "Point") | (geom_types == "MultiPoint"))
+
+    if point_idx.any():
+        point_gdf = _clip_points(gdf[point_idx], poly)
+    else:
+        point_gdf = None
+
+    if poly_idx.any():
+        poly_gdf = _clip_line_poly(gdf[poly_idx], poly)
+    else:
+        poly_gdf = None
+
+    if line_idx.any():
+        line_gdf = _clip_line_poly(gdf[line_idx], poly)
+    else:
+        line_gdf = None
+
+    order = pd.Series(range(len(gdf)), index=gdf.index)
+    concat = pd.concat([point_gdf, line_gdf, poly_gdf])
+
+    polys = ["Polygon", "MultiPolygon"]
+    lines = ["LineString", "MultiLineString", "LinearRing"]
+    points = ["Point", "MultiPoint"]
+
+    # Check that the gdf for multiple geom types (points, lines and/or polys)
+    orig_types_total = sum(
+        [
+            gdf.geom_type.isin(polys).any(),
+            gdf.geom_type.isin(lines).any(),
+            gdf.geom_type.isin(points).any(),
+        ]
+    )
+
+    # Check how many geometry types are in the clipped GeoDataFrame
+    clip_types_total = sum(
+        [
+            concat.geom_type.isin(polys).any(),
+            concat.geom_type.isin(lines).any(),
+            concat.geom_type.isin(points).any(),
+        ]
+    )
+
+    # Check if the clipped geometry contains a geometry collection
+    geometry_collection = (concat.geom_type == "GeometryCollection").any()
+
+    # Check there aren't any new geom types in the clipped GeoDataFrame
+    more_types = orig_types_total < clip_types_total
+
+    if orig_types_total > 1 and keep_geom_type:
+        warnings.warn("keep_geom_type can not be called on a mixed type GeoDataFrame.")
+    elif keep_geom_type and not geometry_collection and not more_types:
+        warnings.warn("keep_geom_type was called when no extra geometry types existed.")
+    elif keep_geom_type and (geometry_collection or more_types):
+        orig_type = gdf.geom_type.iloc[0]
+        if geometry_collection:
+            concat = concat.explode()
+        if orig_type in polys:
+            concat = concat.loc[concat.geom_type.isin(polys)]
+        elif orig_type in lines:
+            concat = concat.loc[concat.geom_type.isin(lines)]
+    elif geometry_collection and not keep_geom_type:
+        warnings.warn(
+            "A geometry collection has been returned. Use .explode() to "
+            "decompose the collection object or keep_geom_type=True to remove "
+            "extra geometry types."
```

Comment:
Do we think this warning (and the one below) is needed? (certainly now geometry collections plot fine) 
Two reasons I am asking: 1) warnings can get annoying if you don't care about the result having other geometry types and 2) if we only have some errors/warnings with `keep_geometry_type=True`, then this whole block that checks the types of the input and resulting gdf can be skipped for the default of `keep_geometry_type=False`